File: java/rooms/RoomFactoryImpl.java
package rooms;

import chest.ChestFactory;
import chest.ChestType;
import logic.GameMaster;
import monsters.CreatureFactory;
import monsters.CreatureType;
import rooms.options.*;

public class RoomFactoryImpl implements RoomFactory {
    private final GameMaster muster;
    private final ChestFactory chestFactory;
    private final CreatureFactory creatureFactory;

    public RoomFactoryImpl(GameMaster muster, ChestFactory chestFactory, CreatureFactory creatureFactory) {
        this.muster = muster;
        this.chestFactory = chestFactory;
        this.creatureFactory = creatureFactory;
    }

    @Override
    public GameMasterable createEmptyRoom(RoomType roomType) {
        return new Room(muster, roomType);
    }

    @Override
    public GameMasterable createRoomWithChest(RoomType roomType, ChestType chestType) {
        GameMasterable baseRoom = createEmptyRoom(roomType);
        return new RoomWithChest(baseRoom,chestFactory.createRandomChest());
    }

    @Override
    public GameMasterable createRoomWithCreature(RoomType roomType, CreatureType creatureType) {
        GameMasterable baseRoom = createEmptyRoom(roomType);
        return new RoomWithCreature(baseRoom,creatureFactory.createCreatureFromType(creatureType));
    }

}

File: php/rooms/RoomFactoryImpl.php
<?php

namespace rooms;

use chest\ChestFactory;
use chest\ChestType;
use logic\GameMaster;
use monsters\CreatureFactory;
use monsters\CreatureType;
use rooms\options\GameMasterable;
use rooms\options\Room;
use rooms\options\RoomWithChest;
use rooms\options\RoomWithCreature;

class RoomFactoryImpl implements RoomFactory {
    private $muster;
    private $chestFactory;
    private $creatureFactory;

    public function __construct(GameMaster $muster, ChestFactory $chestFactory, CreatureFactory $creatureFactory) {
        $this->muster = $muster;
        $this->chestFactory = $chestFactory;
        $this->creatureFactory = $creatureFactory;
    }

    public function createEmptyRoom(RoomType $roomType): GameMasterable {
        return new Room($this->muster, $roomType);
    }

    public function createRoomWithChest(RoomType $roomType, ChestType $chestType): GameMasterable {
        $baseRoom = $this->createEmptyRoom($roomType);
        return new RoomWithChest($baseRoom, $this->chestFactory->createRandomChest());
    }

    public function createRoomWithCreature(RoomType $roomType, CreatureType $creatureType): GameMasterable {
        $baseRoom = $this->createEmptyRoom($roomType);
        return new RoomWithCreature($baseRoom, $this->creatureFactory->createCreatureFromType($creatureType));
    }
}
?>